build-tools/xa-prep-tasks/Xamarin.Android.BuildTools.PrepTasks/GitCommitInfo.cs

```diff
@@ -0,0 +1,322 @@
+				return false;
+			}
+
+			return ParseURL (url.Substring (8, slash - 8), url.Substring (slash + 1), out organization, out repo);
+		}
+
+		bool ParseRelativePathURL (string url, out string organization, out string repo)
+		{
+			return ParseURL (null, url.Substring (6), out organization, out repo);
+		}
+
+		bool ParseURL (string host, string path, out string organization, out string repo)
+		{
+			if (String.IsNullOrEmpty (host))
+				host = "github.com"; // for monodroid, if ever
+
+			string[] parts = path.Split (new [] { '/' }, StringSplitOptions.RemoveEmptyEntries);
+			if (parts.Length < 2) {
+				organization = repo = null;
+				Log.LogError ($"Unknown URL path format '{path}'");
+				return false;
+			}
+
+			if (String.Compare ("github.com", host, StringComparison.Ordinal) == 0) {
+				organization = parts [0];
+				repo = parts [1];
+				if (repo.EndsWith (".git", StringComparison.Ordinal))
+					repo = repo.Substring (0, repo.Length - 4);
+				return true;
+			}
+
+			Log.LogWarning ($"Unknown GIT host '{host}', making assumptions about the URL");
+			organization = parts [0];
+			repo = String.Join ("/", parts, 1, parts.Length - 1);
+
+			return true;
+		}
+
+		bool RunGit (string arguments, List<string> stdoutLines, List<string> stderrLines)
+		{
+			stdoutLines?.Clear ();
+			stderrLines?.Clear ();
+
+			bool canContinue = true;
+			int exitCode = RunCommand (GitPath, arguments, stdoutLines, stderrLines);
+
+			if (exitCode != 0) {
+				canContinue = false;
+				Log.LogError ($"'{GitPath} {arguments}' exited with code {exitCode}");
+			}
+
+			if (stderrLines.Count > 0) {
+				canContinue = false;
+				foreach (string line in stderrLines) {
+					Log.LogError (line);
+				}
+			}
+
+			if (stdoutLines.Count == 0) {
+				canContinue = false;
+				Log.LogError ($"'{GitPath} {arguments}' produced no output");
+			}
+
+			return canContinue;
+		}
+
+		int RunCommand (string commandPath, string arguments, List<string> stdoutLines, List<string> stderrLines)
+		{
+			var si = new ProcessStartInfo (commandPath) {
+				UseShellExecute = false,
+				CreateNoWindow = true,
+			};
+
+			si.WorkingDirectory = WorkingDirectory;
+			si.RedirectStandardOutput = stdoutLines != null;
+			si.RedirectStandardError = stderrLines != null;
+			si.StandardOutputEncoding = Encoding.Default;
+			si.StandardErrorEncoding = Encoding.Default;
+			si.Arguments = arguments;
+			si.EnvironmentVariables.Add ("LC_LANG", "C");
+
+			ManualResetEvent stdout_completed = null;
+			if (!si.RedirectStandardError)
+				si.StandardErrorEncoding = null;
+			else
+				stdout_completed = new ManualResetEvent (false);
+
+			ManualResetEvent stderr_completed = null;
+			if (!si.RedirectStandardOutput)
+				si.StandardOutputEncoding = null;
+			else
+				stderr_completed = new ManualResetEvent (false);
+
+			var p = new Process {
+				StartInfo = si
+			};
+			p.Start ();
+
+			var outputLock = new Object ();
+
+			if (si.RedirectStandardOutput) {
```

Comment:
I think there's a potential ordering issue here.  (In part because `Process` is like THE WORST API in .NET...)

I believe that setting `p.OutputDataReceived` and `p.ErrorDataReceived` needs to be done *before* calling `p.Start()`, to ensure that you don't miss any output.

Reply:
At least, that's what the examples do: https://docs.microsoft.com/sl-si/dotnet/api/system.diagnostics.process.outputdatareceived?view=netframework-3.0

Reply:
The ordering is correct. The collection begins only when you call `Begin*ReadLine()` - and that has to be called after `.Start()` or it crashes. The installer has been using similar code for 7 years without issues (MD too IIRC)